Add support for sqlite loadable extensions

jrandall opened the pull request:
Implements support to author sqlite loadable extensions (https://sqlite.org/loadext.html) using rusqlite, including the ability to write vtables in Rust and use them from other languages or the `sqlite3` CLI. 

I do not appear to be permitted to re-open the original PR #532 so this is resurrecting that PR, while also incorporating improvements made by @gwenn in PR #631 and @phiresky in PR #868.

The two main features that configure this are `loadable_extension` or `loadable_extension_embedded`. The first is used when a rust crate is a standalone loadable extension that itself implements an extension entrypoint (i.e. by implementing an `sqlite3_X_init` function ). The crate can be configured to build as a c lib and then directly loaded into a sqlite that has support for loadable extensions. 

The embedded version is intended if the rust crate implements functionality that is statically linked into a loadable extension implemented outside rust (e.g. in C code). For example, a rust crate using `loadable_extension_embedded` could implement a vtable but not provide a loadable extension entrypoint itself, but rather expects to be embedded within a loadable extension.

The generated sqlite bindings are different when loaded as an extension (note that in C one would need to include `sqlite3ext.h` instead of `sqlite3.h`), so this change requires additional pre-built bindings to be created which have been named with a `-ext.rs` suffix.

jrandall committed:
add support for sqlite loadable extensions

jrandall committed:
install rustfmt in travis ci

jrandall committed:
add rustfmt to appveyor install

jrandall committed:
fix reference to link_lib in vcpkg build

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

# Conflicts:
#	Cargo.toml
#	libsqlite3-sys/Cargo.toml
#	libsqlite3-sys/build.rs
#	src/inner_connection.rs
#	src/lib.rs

gwenn committed:
Remove old bindgens

gwenn committed:
Remove some more old bindgens

gwenn committed:
Remove unused `api_routines_stub` feature

gwenn committed:
Remove unused `vtab_v3` feature

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

gwenn committed:
Revert changes on old bindgens

gwenn committed:
Ignore PATH change on Windows

gwenn committed:
Remove copyright header from build script

gwenn committed:
Generate all three bindgen files

gwenn committed:
Generate new bindgen files for bundled version

gwenn committed:
Remove generate-bindgen-bindings script

If we need to regenerate old bindings,
we can still do it with upgrade.sh

gwenn committed:
Upgrade syn / quote

gwenn committed:
Fix warning after wrong merge

gwenn committed:
Run rustfmt manually only while regenerating bindgen files

gwenn committed:
Fix warnings in build script

gwenn committed:
Remove code specific to SQLITE_DETERMINISTIC

gwenn committed:
Fix build_loadable_extension conditional compilation

gwenn committed:
Remove non_threadsafe feature

gwenn committed:
Make loadable_extension compatible with bundled* features

Like sqlcipher

gwenn committed:
Fix clippy warnings

gwenn committed:
Fix compilation error with loadable_extension feature

gwenn committed:
Revert output mutability

gwenn committed:
Fix warnings

gwenn committed:
Check loadable_extension build

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

# Conflicts:
#	libsqlite3-sys/build.rs

jrandall committed:
eliminate -ext-embed versions of bindgen-bindings

gwenn committed:
Merge pull request #2 from Genomicsplc/pr/loadable-extensions

eliminate -ext-embed versions of bindgen-bindings

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

# Conflicts:
#	src/inner_connection.rs

gwenn committed:
Fix bindgen_bundled_version-ext.rs

gwenn committed:
Add comments

gwenn committed:
`config_log` cannot be used with `loadable_extension`

gwenn committed:
Introduce `to_sqlite_error`

gwenn committed:
Add `dummy-extension` module

gwenn committed:
Fix dummy-extension module

Register a dummy virtual table
Generate a C dynamic library

gwenn committed:
Document and export check macro

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

# Conflicts:
#	.github/workflows/main.yml
#	libsqlite3-sys/build.rs

gwenn committed:
Try to fix CI build

gwenn committed:
Fix dummy-extension

Module reference must be static

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

gwenn committed:
Merge remote-tracking branch 'upstream/master' into pr/loadable-extensions

gwenn committed:
Fix build error

unresolved import `crate::util::alloc`

gwenn committed:
Fix bindgen ext files.

gwenn committed:
Export check macro

gwenn committed:
Fix to_sqlite_error doc

gwenn committed:
Fix dummy extension

gwenn committed:
Fix sqlite3_api_routines size

phiresky committed:
Merge remote-tracking branch 'upstream/master' into gwenntest

phiresky committed:
fix build for loadable_extension

phiresky committed:
update bundled bindgen for loadable_extension

phiresky committed:
add test function to dummy module as well

phiresky committed:
format

phiresky committed:
Merge remote-tracking branch 'upstream/master' into loadable-extensions

jrandall commented:
Please note that the `LICENSE` file in this PR reads as follows:
```
Copyright (c) 2014 John Gallagher <johnkgallagher@gmail.com>
Copyright (c) 2019 Genomics plc <info@genomicsplc.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
```

These are the terms under which we submit the code in this PR that was written by Genomics plc (this was also the case in the original #532 ).

It looks like two other PRs that were based on #532 (#631 and #868) have incorporated the code without following the requirements of the license. I suspect this is unintentional, but I bring it up here in an attempt to prevent it from happening again.

I suspect the authors of #532 and/or #868 probably think that since the `LICENSE` file in master now has a line that reads "Copyright (c) 2014-2020 The rusqlite developers" that is intended to cover this contribution. However, I believe the license terms are very clear that the actual notice is what must be included, so I do not believe this is the case given that this code has been contributed under the terms set out in the `LICENSE` file included in PR #532 (and now again in this one).

As far as I can tell, there is no list of developers included in the rusqlite source code. As such I gather than the intent of the copyright notice naming "The rusqlite developers" is to capture (in a broad sense) all developers who have contributed to the project. If it were me personally, I would be happy to contribute under those terms. However, in this case the work we have contributed is not owned by me but rather by my employer ("Genomics plc") and we have a company policy that work that we contribute to open source projects must be attributed to the company. The usual way to do that is for a copyright statement for "Genomics plc" to be included in every source file where we have made copyrightable contributions. However, we can also accept including the copyright notice in a project-level file such as `LICENSE` where it is clear that intends to cover work throughout the project. 

I can understand if the project does not wish to have a growing list of copyright holders added to the `LICENSE` file, but the terms set out by the owner of the code requires that copyright to be listed somewhere (I could move it back into the source files if that would be preferable).

I could also check with our company legal team if it would be acceptable for our company name ("Genomics plc") to be included in a list of developers/contributors (such as `CONTRIBUTORS`) instead of being directly listed in the `LICENSE` or or source files. I suspect that I might get permission for that if it is clear that "The rusqlite developers"  includes entities listed in the contributors file.

Which approach would be preferable?

jrandall commented on src/inner_connection.rs:
```diff
@@ -14,7 +14,6 @@ use crate::error::{error_from_handle, error_from_sqlite_code, Error};
 use crate::raw_statement::RawStatement;
 use crate::statement::Statement;
 use crate::unlock_notify;
-use crate::version::version_number;
```
Due to conditional compilation, `version_number` is not referenced when both `bundled` and `loadable_extension`, so its use here triggers a lint. It has been moved into the blocks where it is actually used.

jrandall commented on src/inner_connection.rs:
```diff
@@ -4,7 +4,7 @@ use std::os::raw::{c_char, c_int};
 use std::path::Path;
 use std::ptr;
 use std::str;
-use std::sync::atomic::{AtomicBool, Ordering};
```
Same situation with `Ordering` as with `version_number` https://github.com/rusqlite/rusqlite/pull/910/files#r588409133

gwenn commented:
See https://github.com/rusqlite/rusqlite/pull/631#discussion_r404001452

thomcc commented:
So, for the patch in the branch that hasn't landed yet, I don't have strong feelings about what to do about it, although it's clear removing it does violate the "The above copyright notice and this permission notice shall be included" bit of the license. I agree that reverting or rebasing the commit away is probably the best action there.

---

More generally though, this is a policy question.

Currently, we no longer list individual contributor names in authors/license, as we don't take copyright of contributions (doing so would require a CLA), and "rusqlite developers/authors/contributors" most accurately reflects this fact. My understanding is that you're still totally free to assign copyright to your employer as you otherwise would be.

It's also much easier to maintain, doesn't leave people out, imply anything about the value of some contributions or others. More specifically for this case, it doesn't imply anything about the relative control an entity has over the project (that is: it's somewhat reasonable for a third-party to assume that if the only two entries named in the license were a nebulous "Rusqlite Authors" and $SomeCompany, that the company is probably the main or a substantial stakeholder in the project).

(If it helps, this is not the first set of contributions that were funded by company time. I used to be paid to work on rusqlite for Mozilla, and I'm sure others have submitted patches on behalf of their employers as well)

---

Concretely: I respect your company's wish to include their name in all license files of OSS projects that they spend money (via their employees time) to maintain, but (while I'd rather it not come to this) we're also free just to not accept contributions that come with strings we don't like attached.

Sorry if that seems curt, but... you are sort of demanding that we change how the repository operates — you aren't going to find a lot of open source projects that are thrilled with that — most projects above a certain size have a hard no for that kind of thing (and if it were just me on rusqlite, that would probably be my stance as well).

That said, this would be nice to come to consensus here — @gwenn has put in a lot of effort already, and there are definitely use cases this enables.

jrandall commented:
> So, for the patch in the branch that hasn't landed yet, I don't have strong feelings about what to do about it, although it's clear removing it does violate the "The above copyright notice and this permission notice shall be included" bit of the license. I agree that reverting or rebasing the commit away is probably the best action there.
> 
> More generally though, this is a policy question.
> 
> Currently, we no longer list individual contributor names in authors/license, as we don't take copyright of contributions (doing so would require a CLA), and "rusqlite developers/authors/contributors" most accurately reflects this fact. My understanding is that you're still totally free to assign copyright to your employer as you otherwise would be.
> 
> It's also much easier to maintain, doesn't leave people out, imply anything about the value of some contributions or others. More specifically for this case, it doesn't imply anything about the relative control an entity has over the project (that is: it's somewhat reasonable for a third-party to assume that if the only two entries named in the license were a nebulous "Rusqlite Authors" and $SomeCompany, that the company is probably the main or a substantial stakeholder in the project).
> 
> (If it helps, this is not the first set of contributions that were funded by company time. I used to be paid to work on rusqlite for Mozilla, and I'm sure others have submitted patches on behalf of their employers as well)
> 
> Concretely: I respect your company's wish to include their name in all license files of OSS projects that they spend money (via their employees time) to maintain, but (while I'd rather it not come to this) we're also free just to not accept contributions that come with strings we don't like attached.
> 
> Sorry if that seems curt, but... you are sort of demanding that we change how the repository operates — you aren't going to find a lot of open source projects that are thrilled with that — most projects above a certain size have a hard no for that kind of thing (and if it were just me on rusqlite, that would probably be my stance as well).
> 
> That said, this would be nice to come to consensus here — @gwenn has put in a lot of effort already, and there are definitely use cases this enables.

I agree you are free not to accept the contribution if you don't want to provide a place for specific copyright notices. I will check with the legal team as to whether we are happy to be an unnamed part of "The rusqlite developers" but that has not been the case in previous situations. I'm really not trying to be a jerk about this - it's just that I have permission to distribute this work which is owned by my employer only under the permitted terms.

We have not run into this problem with any other open source project - there have been situations where the project wanted to change where our notice was included (for example, as apparently happened here in being asked to move the notice from the source file to a repo-level `LICENSE` file) but in every case thus far the project we have contributed to has had an acceptable place for our copyright notice to be represented. 

I do appreciate the situation would be imbalanced to simply include our name alongside "The rusqlite developers" in the LICENSE file. I have sought and received permission from our legal team to have our name not be mentioned in the LICENSE itself but only in an explanatory note such as a `CONTRIBUTORS` file, but I don't think we will be able to contribute without our name appearing somewhere.

My suggestion for a compromise would be to add a CONTRIBUTORS file that says something like this:

```
"The rusqlite developers" referred to in the LICENSE copyright notice includes all individual developers who have contributed code to the project as well as the following organisations who retain copyright in portions of the code (this list is not intended to be complete):
 - Genomics plc
```

If there are any other organisations known to have a copyright in code contributed in the project, their names could be included there as well to make our inclusion seem less prominent. 

I'm afraid this is as far as our policy can stretch. I respect the work you do as maintainers and you are certainly free to reject our code if we cannot come to terms on this. In that case I sincerely apologise to the developers who have also put substantial work into the effort to get this functionality into a mergeable state. If the project cannot accommodate any form of company attribution, we will not be able to contribute code to it and will have to either maintain our fork in private or publicly fork the project and publish it under a different name.

jrandall commented:
I have now moved the "Genomics plc" attribution out of `LICENSE` and into an explanatory note in `CONTRIBUTORS.md`

jrandall commented on src/trace.rs:
```diff
@@ -8,8 +8,7 @@ use std::ptr;
 use std::time::Duration;
 
 use super::ffi;
-use crate::error::error_from_sqlite_code;
```
This use has been moved into the config_log function since it is now conditionally compiled only when not a loadable extension.

jrandall commented on src/vtab/mod.rs:
```diff
@@ -1063,12 +1044,6 @@ unsafe fn result_error<T>(ctx: *mut ffi::sqlite3_context, result: Result<T>) ->
     }
 }
 
-// Space to hold this string must be obtained
```
`alloc` function moved to util module as it is now used by both vtab and error modules.

phiresky commented:
> My suggestion for a compromise would be to add a CONTRIBUTORS file [...]

That looks like a great solution! I've closed the other PR. Can we merge this now? :) 

Personally I can say I don't really care who exactly owns the copyright, as long as the terms of the license stay the same. 

I think many projects also have the same thing called an [AUTHORS](https://opensource.google/docs/releasing/authors/) file, that's sometimes generated from the commit logs with added organisations when the individual contributors don't themselves own the copyright. Linux has a [CREDITS](https://github.com/torvalds/linux/blob/master/CREDITS) file where people and organisations add themselves with some more details.

jrandall commented:
I'm working on fixing the test failures. It looks like there are some issues with exotic combinations of feature flags.

jrandall commented:
> > My suggestion for a compromise would be to add a CONTRIBUTORS file [...]
> 
> That looks like a great solution! I've closed the other PR. Can we merge this now? :)
> 
> Personally I can say I don't really care who exactly owns the copyright, as long as the terms of the license stay the same.
> 
> I think many projects also have the same thing called an [AUTHORS](https://opensource.google/docs/releasing/authors/) file, that's sometimes generated from the commit logs with added organisations when the individual contributors don't themselves own the copyright. Linux has a [CREDITS](https://github.com/torvalds/linux/blob/master/CREDITS) file where people and organisations add themselves with some more details.

The approach Google uses with the `AUTHORS` file seems very clear and would also work for us. It seems that our legal team is willing to be flexible in terms of going along with whatever specific mechanism the rusqlite maintainers would like to use to represent the attribution, but we cannot contribute significant work without the company being mentioned by name somewhere.

gwenn reviewed (changes requested):
Commit history has been lost !!!
I am afraid you will have to restart from @phiresky branch.

jrandall commented on dummy-extension/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "dummy-extension"
+version = "0.0.1"
+authors = ["John Gallagher <jgallagher@bignerdranch.com>"]
```
@gwenn did you write the dummy-extension? If so, shall we change this line to "The rusqlite developers" ?

gwenn replied:
I guess that this file was created before `rusqlite` organisation and not yesterday as suggested by your branch history...
See https://github.com/rusqlite/rusqlite/pull/910#pullrequestreview-605855451

jrandall replied:
Got it - I did start from the phiresky branch but it was a bit of rebasing hell to get all of those commits up to date with master. I ended up squashing everything while I sorted out line by line every single change with respect to master, but I've now replayed my changes on top of the 868 commit history, and indeed it appears I made a mistake during the squashing in which the original authors line was there instead of "The rusqlite developers".

jrandall committed:
Fix up loadable extensions work

jrandall commented:
> Commit history has been lost !!!
> I am afraid you will have to restart from @phiresky branch.

History has been restored. Apologies - I had squashed it all temporarily while working with it locally and I forgot to restore the history before I opened the PR.  We have a lot of private work built on top of this and the commit history does not rebase cleanly, so it was easier to work with shipping a single commit around while doing local integration/testing work.

jrandall committed:
Merge branch 'master' into loadable-extensions

jrandall committed:
improve loadable_extension integration test and add loadable_extension_embedded integration test

trevyn commented:
The "dummy" extensions in the code are just example extensions, right? Would it make sense to rename all instances of "dummy" to "example", to avoid any future... inclusivity concerns?

jrandall commented:
> The "dummy" extensions in the code are just example extensions, right? Would it make sense to rename all instances of "dummy" to "example" or "demo" or similar, to avoid any future... inclusivity concerns?

I agree that is both more clear and avoids potential issues surrounding the "dummy" term - @gwenn are you happy for me to do the rename from "dummy" to "example"?

jrandall committed:
Merge tag 'v0.25.0' into loadable-extensions

jrandall committed:
Merge tag 'v0.26.2' into loadable-extensions

rusqlite-v0.26.2, libsqlite3-sys-0.23.2

jrandall committed:
rename dummy-extension and dummy-embedded-extension to example-extension and example-embedded-extension

jrandall committed:
fix issues resulting from merge conflict resolution in libsqlite3-sys/build.rs

jrandall committed:
fix issue with header_path in libsqlite3-sys/build.rs

correct clippy::needless-borrow issue in to_sqlite_error function in src/error.rs

jrandall commented:
> The "dummy" extensions in the code are just example extensions, right? Would it make sense to rename all instances of "dummy" to "example" or "demo" or similar, to avoid any future... inclusivity concerns?

I've made the change from "dummy" to "example"

jrandall committed:
don't specify version in example-*/Cargo.toml so that they can be built against the current source tree

jrandall committed:
cargo fmt

jrandall commented:
@gwenn I addressed the review changes you requested on 7th March - could you please re-review? I've now also updated against the default branch for v0.26.2 and fixed all issues with the tests.

gwenn commented:
> @gwenn I addressed the review changes you requested on 7th March - could you please re-review? I've now also updated against the default branch for v0.26.2 and fixed all issues with the tests.

We should review all comments from @phiresky PR...

jrandall commented on libsqlite3-sys/src/lib.rs:
```diff
@@ -56,6 +56,25 @@ pub enum Limit {
+#[cfg(feature = "loadable_extension_embedded")]
+// bindings were built with loadable_extension_embedded:
+// define sqlite3_api as an extern since this code will be embedded
+// within a loadable extension that defines and exports this itself
+extern "C" {
+    pub static mut sqlite3_api: *mut sqlite3_api_routines;
+}
```
@thomcc commented on this in the now closed PR: https://github.com/rusqlite/rusqlite/pull/868#discussion_r546279767
"This should probably be an AtomicPtr, and have a Release store and Acquire loads. Otherwise it's a data race."

jrandall replied:
This was discussed against the earlier PR: https://github.com/rusqlite/rusqlite/pull/631#discussion_r416921802

jrandall replied:
I will attempt the proposed addition of `sqlite_extension_init` that I suggested in https://github.com/rusqlite/rusqlite/pull/631#discussion_r416921802

jrandall replied:
This has now been implemented as `loadable_extension_init` and `loadable_extension_embedded_init`. 

@thomcc are you happy with this?

thomcc replied:
It seems like you use a once for the write and not for the reads. This is not enough for happens-before. You need an acquire on the reader threads, or for them to attempt to run the `Once` too (which they can't, because they don't the ptr to initialize. So it has a data race.

Instead, storing it with an `AtomicPtr` with a `Release` store, and reading it with an `Acquire` load would be better.

But I think this should be done fairly differently, as mentioned.

gwenn replied:
@thomcc Something likes [that](https://gist.github.com/gwenn/32beb99dec960f73e025e8b322199812) ?

thomcc replied:
You need at least a Release store (and Acquire loads) for proper happens-before.

gwenn replied:
Thanks !
And another question: how to handle [va_list](https://github.com/rusqlite/rusqlite/blob/master/libsqlite3-sys/sqlite3/sqlite3ext.h#L160) ?
I guess we cannot remove this field (`vmprintf`) from the generated bindings (like we currently do for the associated function `sqlite3_vmprintf`).
I've tried to use bindgen [opaque_type("(?i).*va_list.*")](https://docs.rs/bindgen/0.66.1/bindgen/struct.Builder.html#method.opaque_type) without success.
So temporary, I've manually replaced `va_list` by `void*`.

gwenn replied:
@thomcc I manage to generate a `bindgen_bundled_version_ext.rs` with [this](https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs) code.
```rust
...
static __SQLITE3_LAST_INSERT_ROWID: ::std::sync::atomic::AtomicPtr<
    unsafe extern "C" fn(arg1: *mut sqlite3) -> sqlite_int64,
> = ::std::sync::atomic::AtomicPtr::new(std::ptr::null_mut());
pub unsafe fn sqlite3_last_insert_rowid(arg1: *mut sqlite3) -> sqlite_int64 {
    let fun = __SQLITE3_LAST_INSERT_ROWID.load(::std::sync::atomic::Ordering::Acquire);
    assert!(! fun.is_null(), "SQLite API not initialized");
    (*fun)(arg1)
}
...

/// Like SQLITE_EXTENSION_INIT2 macro
pub unsafe fn rusqlite_extension_init2(p_api: *mut sqlite3_api_routines) {
    assert!(! p_api.is_null(), "Invalid sqlite3_api_routines pointer");
...
    __SQLITE3_LAST_INSERT_ROWID
        .store(
            &mut (*p_api).last_insert_rowid.unwrap(),
            ::std::sync::atomic::Ordering::Release,
        );
...
}
```
Could you please tell me if this is the right way ?
Thanks.

thomcc replied:
I think instead of `&mut (*p_api).last_insert_rowid.unwrap()` you should store the function address directly, since the result of `.last_insert_rowid.unwrap()` is a temporary, and you're storing a pointer to the result of a temporary. I'm a little surprised it works.

There are a few ways you could fix it but the way with the best performance would be https://gist.github.com/thomcc/242dcdcc3fa840567dd4ea7c75ee896c probably. Let me know if that makes sense to you.

I don't think there's anything we can do for functions with `va_list` or `...`

gwenn replied:
@thomcc 
(a) What about the [atomic](https://crates.io/crates/atomic) crate which gives [this](https://gist.github.com/gwenn/32beb99dec960f73e025e8b322199812) ?
```rust
static _sqlite3_db_config: Atomic<
    Option<unsafe extern "C" fn(arg2: ::std::os::raw::c_int, ...) -> ::std::os::raw::c_int>,
> = Atomic::new(None);
```
No need to transmute and is lock free:
```rust
    println!(
        "is_lock_free: {}",
        Atomic::<
            Option<unsafe extern "C" fn(arg2: ::std::os::raw::c_int, ...) -> ::std::os::raw::c_int>,
        >::is_lock_free()
    ); // is_lock_free: true
```

(b) And for `va_list`, how can we make the generated bindings "portable" for:
```c
struct sqlite3_api_routines {
...
  char *(*vmprintf)(const char*,va_list);
...
  char *(*xvsnprintf)(int,char*,const char*,va_list);
...
  void (*str_vappendf)(sqlite3_str*, const char *zFormat, va_list);
...
};
```
?
Thanks.

thomcc replied:
The atomic crate looks fine if used as you describe. I think the other stuff is not possible to define correctly in Rust at the moment (prior to `core::ffi::VaList` stabilizing), but you could just the function pointers filled out with `*const ()` or something, as a placeholder. There's no real reason for people to use those functions in Rust anyway.

gwenn replied:
Ok,
I am able to do a:
```sh
rusqlite % cargo check --features "modern_sqlite,loadable_extension,vtab,trace"
```
by manually fixing the signature of `sqlite3_db_config` and `sqlite3_log` (see https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs-L135-L152).
Then replacing locally the default bindings by the one generated by https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs :
```sh
rusqlite % cp ../sqlite-ext/bindgen_bundled_version_ext.rs libsqlite3-sys/sqlite3/bindgen_bundled_version.rs
```
And applying this minimal patch: https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-rusqlite-patch.
So the last remaining step is fixing libsqlite3-sys/build.rs by integrating https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs...

jrandall commented on example-extension/src/lib.rs:
```diff
@@ -0,0 +1,118 @@
+};
+use rusqlite::{to_sqlite_error, Connection, Result};
+
+#[allow(clippy::not_unsafe_ptr_arg_deref)]
+#[no_mangle]
+pub extern "C" fn sqlite3_extension_init(
```
Commented on by @thomcc in #631 https://github.com/rusqlite/rusqlite/pull/631#discussion_r404001494
"
This probably should be `unsafe fn` though. It both writes to a static mut without synchronization and dereferences the raw pointers passed in as arguments.
"

jrandall replied:
fixed

jrandall committed:
remove extraneous cmoment line in libsqlite3-sys/build.rs

jrandall commented on libsqlite3-sys/build.rs:
```diff
@@ -593,47 +712,377 @@ mod bindings {
+            let mut_c_int_type: syn::Type = syn::parse2(quote!(*mut ::std::os::raw::c_int))
+                .expect("failed to parse mutable c_int reference");
+            // until rust c_variadic support exists, we can't
+            // transparently wrap variadic api functions.
+            // generate specific set of args in place of
+            // variadic for each function we care about.
+            match api_fn_name {
+                "sqlite3_db_config" => {
+                    // https://sqlite.org/c3ref/c_dbconfig_defensive.html
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_db_config_constchar".to_string(),
+                        generate_varargs_input_idents(field_ident, &bare_fn, &[&const_c_char_type]),
+                    ); // used for SQLITE_DBCONFIG_MAINDBNAME
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_db_config_void_int_mutint".to_string(),
+                        generate_varargs_input_idents(
+                            field_ident,
+                            &bare_fn,
+                            &[&mut_void_type, &c_int_type, &mut_c_int_type],
+                        ),
+                    ); // used for SQLITE_DBCONFIG_LOOKASIDE
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_db_config_int_mutint".to_string(),
+                        generate_varargs_input_idents(
+                            field_ident,
+                            &bare_fn,
+                            &[&c_int_type, &mut_c_int_type],
+                        ),
+                    ); // used for all other configuration verbs
+                }
+                "sqlite3_vtab_config" => {
+                    // https://sqlite.org/c3ref/c_vtab_constraint_support.html
+                    wrapper_fn_inputs_map.insert(
+                        "sqlite3_vtab_config_int".to_string(),
+                        generate_varargs_input_idents(field_ident, &bare_fn, &[&c_int_type]),
+                    ); // used for SQLITE_VTAB_CONSTRAINT_SUPPORT
+                }
+                _ => {}
+            };
+        }
```
An older version of this code was commented on by @thomcc in https://github.com/rusqlite/rusqlite/pull/631/files/cf658af142399be749bacb0387d54ce836b85f1a#r404006799
"
This isn't really safe. Vararg calling convention isn't guaranteed to be the same as the fixed arg calling convention on all platforms / compilers (https://www.gnu.org/software/libc/manual/html_node/Calling-Variadics.html mentions this).

That said I guess it should be fine in practice, but still is worrisome...
"

jrandall replied:
Don't currently have a proposal to address this. At the moment we are only implementing calls to a small number of variadic API functions (that I am actually using). 

@thomcc would you be happier if inclusion of these bindings was behind a feature? Something like `experimental_variadic_support`?

thomcc replied:
Hmm, I'm not sure if I wrote this elsewhere — worth noting that aarch64 darwin (so ios/tvos/macos m1) is a target where the varargs and fixargs calling conventions are different, so it's not really a theoretical concern.

That said I need to look over the code again, I'm unsure why it can't work as varargs.

jrandall commented on libsqlite3-sys/build.rs:
```diff
@@ -593,47 +712,377 @@ mod bindings {
+
+            // generate wrapper and return it as a string
+            let wrapper_tokens = quote! {
+                pub unsafe fn #api_fn_ident(#api_fn_inputs) #api_fn_output {
+                    if sqlite3_api.is_null() {
+                        panic!("sqlite3_api is null");
+                    }
+                    ((*sqlite3_api).#field_ident
+                        .expect(stringify!("sqlite3_api contains null pointer for ", #field_name, " function")))(
+                            #(#api_fn_input_idents),*
+                    )
+                }
```
Previously commented on (by me) in https://github.com/rusqlite/rusqlite/pull/631#issuecomment-620874452
"
Extension API function wrappers will be generated for all API functions in the sqlite bindings present at build time. It seems there is a trend towards encouraging building rusqlite against a very recent version of sqlite when any non-ancient features are requested (gating them all with the `modern_sqlite` feature - see #639 ). If this is the case, vtable authors will have little choice but to disable the version check in order to make it possible for end users to use their extension in a variety of environments. This presents a serious safety risk as an attempt to call an API function from a newer version (i.e. from the wrapper at

https://github.com/rusqlite/rusqlite/blob/cf658af142399be749bacb0387d54ce836b85f1a/libsqlite3-sys/build.rs#L638

) would result in reading past the end of that struct into memory we don't own, and then executing what we find at the target of a pointer we find there.

If we are going to encourage people to disable the version check by making it difficult to build against a minimum required version of sqlite, then we should probably protect the wrappers by including a runtime version check in them (probably using Once) and refusing to attempt to execute any belonging to a version newer than the runtime environment supports.
"

And https://github.com/rusqlite/rusqlite/pull/631#issuecomment-620877082
"
Version requirements for the API routines are documented in comments in `sqlite3ext.h` (e.g. https://github.com/rusqlite/rusqlite/blob/master/libsqlite3-sys/sqlite3/sqlite3ext.h#L287) but I am not sure what the best way would be to capture this and pass it through to the wrapper function generator so we can populate a required version test in each of them. 
"

jrandall replied:
Runtime version checking has now been implemented by generating a separate set of rust-bindgen bindings from `sqlite3ext.h` including comments and parsing the `sqlite_api_routine` fields along with the inline version comments that indicate minimum versions for each section of added fields.

jrandall commented:
> > @gwenn I addressed the review changes you requested on 7th March - could you please re-review? I've now also updated against the default branch for v0.26.2 and fixed all issues with the tests.
> 
> We should review all comment(s) from @phiresky PR: for example, [#868 (review)](https://github.com/rusqlite/rusqlite/pull/868#pullrequestreview-555966681)...

Good point - there were actually a whole bunch of earlier ones from #631 as well. I've gone through all of the comments in #532 #631 and #868 and referenced any extant threads here, along with a cross-reference link to the new thread. 

There are a few outstanding issues that need to be addressed.

jrandall committed:
add runtime version checks to loadable extension wrappers, improve safety, and add explicit entry points for both embedded and non-embedded loadable extensions

jrandall committed:
cargo fmt

jrandall committed:
add safety docs to unsafe functions in example-extension and example-embedded-extension

jrandall committed:
cargo fmt

thomcc commented:
First thoughts below (actual review later — It's late and I got sidetracked).

I am genuinely concerned about the safety issue here with versions... I don't think requiring the version match the generated bindings is okay, since (for example), I'm planning on getting rid of as many of the bindings files that aren't prebundled as I can (buildtime-bindgen will still exist) when I have time[^1]. This fits with what the other libsqlite3-sys using crates have been asking for a while, and also will simplify things and make fewer changes breaking for people who separately compile sqlite.

[^1]: I need to do a writeup of the plan first though, since this is a big change and I want to be sure that if I axe someone's key feature, they at least have time to say something.

Anyway, a way forward here. Hm. I have some code where I tried to parse the C sqlite3ext.h to get the versions, which worked fine (the original motivation for it was unrelated to extensions, actually...)

Okay, I dug out the code. It's here, and I extended it to do minimal parsing of the `sqlite3_api_routines` struct in order to determine the offset of each function: https://gist.github.com/thomcc/319bffa19575d2eab95a0eae5a0c1dca. Now that I've written it, I'm not sure that I think this is a good idea[^2] though.

[^2]: I was intending on suggesting that we use this to store the size/max offset the struct should be for that version (populating a global during extension init), and then use memoffset::offset_of to determine whether or not it woudl be okay to access the field whenever we call. I don't like this, the more I think about it, although it would work.

A better way is probably to store and read into the functions themselves, rather the the API. This would also give us a path forward to eventually supporting `dlopen`/`dlsym`.

Specifcally, rather than a single `AtomicPtr<sqlite3_api_routines>`, we'd effectively have one `Atomic<Option<unsafe extern "C" fn(...)>>`[^3] per field. Then, when initializing the extension API, we'd call `libversion_number` from api pointer, and use that to figure out which fields should be valid, and populate the supported function pointers.

[^3]: Well, conceptually anyway. In reality `Atomic<Option<fn>>` is not possible, so instead you need to use `AtomicPtr<()>` and transmute the result of the load to `Option<unsafe extern "C" fn(...)>` — conceptually it's basically the same though.

I like this approach more, and was kind of planning on doing it anyway for dlopen/dlsym support. It should be easier after I overhaul libsqlite3-sys, though.

P.S. So, I hate to be that guy and ask, but is this the PR that had all the licensing headaches? My feeling there hasn't really changed.

thomcc commented on libsqlite3-sys/build.rs:
```diff
@@ -593,47 +778,572 @@ mod bindings {
+                pub unsafe fn #api_fn_ident(#api_fn_inputs) #api_fn_output {
+                    let p_api = loadable_extension_sqlite3_api();
+                    #api_version_check_tokens
+                    ((*p_api).#field_ident
```
Watch out: if `sqlite3_api_routines` is actually smaller than we think, then this becomes UB — it creates an intermediate reference, and references are marked as dereferencable (in llvm) for the whole size of the struct.

Instead you have to use `core::ptr::addr_of!((*p).field).read()`, for example.

jrandall replied:
My intent was that the code inserted by #api_version_check_tokens applies the runtime version check and prevents the next line from being executed if `sqlite3_api_routines` is smaller at runtime than at compile time.

I'm not extremely well versed in this part of Rust. I guess the undefined behaviour is because we are never allowed to make a reference (not a raw pointer, but a reference) to uninitialized data even if we don't execute it, because the compiler may rearrange execution order or something like that? 

I'll make the change to raw pointers using `addr_of` and `read` - thanks for the suggestion!

jrandall commented:
> P.S. So, I hate to be that guy and ask, but is this the PR that had all the licensing headaches? My feeling there hasn't really changed.

Yes, it was. My employer owns the original code, and we had originally contributed it with a copyright statement for the company in the source files in which we made contributions (this is the company's usual policy and was also consistent with the state of copyright in the project at the time). Since opening the first PR on this (#532), the project changed its policy and rewrote all of the copyright statements to reference "The rusqlite developers". That unfortunately (and unintentionally as far as I can tell) resulted in all of my employer's copyright statements being removed when the default branch was merged into one of the subsequent PRs. 

I subsequently went to my employer's legal team and got permission to contribute code with copyright statements attributed to "The rusqlite developers" and after some internal debate I was able to get them to agree to that with the provision that our contributions be attributed in an appropriate place such as in documentation or a CONTRIBUTORS file. 

The current PR includes a (new) `CONTRIBUTORS.md` that would satisfy that requirement (https://github.com/rusqlite/rusqlite/blob/069fc87fa12d5ddd220d1550618f0767ec403b39/CONTRIBUTORS.md). However, we are willing to be as flexible as we can be in terms of fitting in with how the project would like to recognize non-individual contributors. If you do not agree that the proposed `CONTRIBUTORS.md` is the right solution, then please suggest an alternative.

It is my understanding that the intent is that the use of "The rusqlite developers" is intended to recognize (generally) all individuals who have contributed to the project, and that many such contributors can be identified through the git commit history, but in this case that would recognize me rather than my employer, and it is my employer who owns the work.

Will the current proposed `CONTRIBUTORS.md` file work for the project? If so, it might be nice to add a few more organizational contributors to it, both to recognise the support they have given the project and also so that it doesn't look like my employer is a more major contributor than they are.

If this won't work, do you have an alternative to suggest?

jrandall commented:
> I am genuinely concerned about the safety issue here with versions... I don't think requiring the version match the generated bindings is okay, since (for example), I'm planning on getting rid of as many of the bindings files that aren't prebundled as I can (buildtime-bindgen will still exist) when I have time[1](#user-content-fn-1-909adde11c647eb66a026fd1ab19894e). This fits with what the other libsqlite3-sys using crates have been asking for a while, and also will simplify things and make fewer changes breaking for people who separately compile sqlite.
> 
> Anyway, a way forward here. Hm. I have some code where I tried to parse the C sqlite3ext.h to get the versions, which worked fine (the original motivation for it was unrelated to extensions, actually...)

@thomcc I am not sure if you saw yet that I have implemented this in the current branch. The build script for libsqlite3-sys includes code to track the version comments in the sqlite headers and set a version check Option for each API function. 

Parse rust-bindgen bindings (generated with comments from `sqlite3ext.h`): https://github.com/rusqlite/rusqlite/blob/069fc87fa12d5ddd220d1550618f0767ec403b39/libsqlite3-sys/build.rs#L968-L1013

Generate runtime version checks for all API routines after the base set that are assumed to work in any version: https://github.com/rusqlite/rusqlite/blob/069fc87fa12d5ddd220d1550618f0767ec403b39/libsqlite3-sys/build.rs#L1279-L1332

jrandall commented:
> A better way is probably to store and read into the functions themselves, rather the the API. This would also give us a path forward to eventually supporting `dlopen`/`dlsym`.
> 
> Specifcally, rather than a single `AtomicPtr<sqlite3_api_routines>`, we'd effectively have one `Atomic<Option<unsafe extern "C" fn(...)>>`[3](#user-content-fn-3-909adde11c647eb66a026fd1ab19894e) per field. Then, when initializing the extension API, we'd call `libversion_number` from api pointer, and use that to figure out which fields should be valid, and populate the supported function pointers.
> 
> I like this approach more, and was kind of planning on doing it anyway for dlopen/dlsym support. It should be easier after I overhaul libsqlite3-sys, though.

I think I see what you are suggesting here, and the new `loadable_extension_init` and `loadable_extension_embedded_init` functions provide the extension point where this could be achieved. 

Your approach would have the benefit of doing this version check once at initialization rather than (potentially) on every call to every API routine. I had thought the current approach could be extended by memoizing (really just Once-ing) the runtime checks for each function, which could also address those concerns (probably with about the same overhead as doing the runtime check against the Option). 

I had also thought about extending the bindings to provide something like an `*_is_supported` function for each of the routines, so that consumers of the library can anticipate and avoid the panic that would otherwise be caused by the runtime checks (i.e. by having some fallback behaviour when certain routines are not available). However, it was starting to feel like there is too much logic being written into the bindings here, and I'd like to be able to push a 

Your proposed interface would address that cleanly if we provide the struct representing the API routines to consumers so that they can check against the Option for the routine of interest prior to invoking it. 

I guess what I'm not clear on here is how we would provide this to consumers of libsqlite3-sys such as rusqlite -- would you propose that the main interface that libsqlite3-sys provides to its consumers (including rusqlite) be refactored so that all accesses to `sqlite3_*` methods go through an API methods struct such as the one you suggest? I guess that would have the possible benefit of decoupling the sqlite3 API routines that the libsqlite3-sys rust library is aware exist from those that are acutally present in the version of sqlite3 that its bindings are built against. Even when not building an extension, an entrypoint could populate the API routines structure when statically compiled, such that those that are present in the sqlite it is built against are `Some(fn ... )` and newer ones that aren't supported are `None`. This would push the responsibility for adapting to different runtime versions of sqlite to the consumers of libsqlite3-sys, but it would also obviously give them the responsibility for handling situations in which certain routines are not available appropriately. 

Or, alternatively, would we continue to offer (unwrapped) `sqlite3_*` methods to consumers of `libsqlite3-sys` but back them using the proposed methods structure in certain situations, with a default way of handling routines (that would probably be to handle a `None` invocation via a panic?) but also offer the wrapped API routines Options interface as an alternative way for consumers to access API routines when they are able to provide fallback handling in the case of API routines that are None?

jrandall committed:
fix doctests in src/loadable_extension.rs

jrandall committed:
access sqlite3_api_routines fields in generated wrapper functions using addr_of! and read to avoid undefined behaviour when runtime sqlite3_api_routines is shorter than expected at compile time

anacrolix commented:
It would be great to get this in, it's being used to good effect (or some variant) with https://crates.io/crates/rusqlite-le and https://github.com/phiresky/sqlite-zstd.

nathaniel-daniel commented:
Fairly certain that rusqlite fork is based on a PR that does not comply its creator's license.

thomcc commented:
I have a patch that does this in a different way, I'll try to resurrect it and finish it up.

avinashkurup commented:
> Implements support to author sqlite loadable extensions (https://sqlite.org/loadext.html) using rusqlite, including the ability to write vtables in Rust and use them from other languages or the `sqlite3` CLI.
> 
> I do not appear to be permitted to re-open the original PR #532 so this is resurrecting that PR, while also incorporating improvements made by @gwenn in PR #631 and @phiresky in PR #868.
> 
> The two main features that configure this are `loadable_extension` or `loadable_extension_embedded`. The first is used when a rust crate is a standalone loadable extension that itself implements an extension entrypoint (i.e. by implementing an `sqlite3_X_init` function ). The crate can be configured to build as a c lib and then directly loaded into a sqlite that has support for loadable extensions.
> 
> The embedded version is intended if the rust crate implements functionality that is statically linked into a loadable extension implemented outside rust (e.g. in C code). For example, a rust crate using `loadable_extension_embedded` could implement a vtable but not provide a loadable extension entrypoint itself, but rather expects to be embedded within a loadable extension.
> 
> The generated sqlite bindings are different when loaded as an extension (note that in C one would need to include `sqlite3ext.h` instead of `sqlite3.h`), so this change requires additional pre-built bindings to be created which have been named with a `-ext.rs` suffix.

Hello,

I am reaching out to inquire about a specific implementation aspect of the `sqlite-html` extension, particularly in the context of its integration with rusqlite.

**Background:**
I came across a concept in Rust where a crate can implement functionality that is statically linked into a loadable extension implemented outside of Rust (e.g., in C code). Specifically, this approach is used in scenarios where the Rust crate, using something like `loadable_extension_embedded`, implements a virtual table (vtable) but does not provide a loadable extension entry point itself. Instead, it expects to be embedded within a loadable extension.

**Query:**
I am curious if a similar approach is possible or already implemented with the `sqlite-html` extension. The core of my question is:

- Can the `sqlite-html` extension, which is primarily in Go, be integrated with rusqlite in such a way that html-based functionality is embedded within rusqlite?
- If so, are there any examples or documentation available that demonstrate this integration, particularly with the use of `loadable_extension_embedded` or a similar method?

This information would be greatly beneficial for a project I am working on, where integrating Rust capabilities with the `sqlite-html` extension could significantly enhance our application.

Thank you for your time and assistance. I look forward to your response and any insights you can provide on this matter.

Best regards,
Avinash.

gwenn commented:
@avinashkurup See #1362
I am not sure that you need it but "statically linked" is not supported yet.

y2kbugger commented:
New to rust and kinda confused about all the forks out there for this mentioned in blogs about extending sqlite with rust. 

if you can hook an extension by just defining a function, what is the point of this pr?
```
/// SQLite extension entry point, named based on the lib filename
#[unsafe(no_mangle)]
pub extern "C" fn sqlite3_sqliteinstrumentrs_init(
    _db: *mut sqlite3,
    pz_err_msg: *mut *mut c_char,
    _p_api: *const sqlite3_api_routines,
) -> c_int {
...
}
```
